include/drivers/pinctrl.h

```diff
@@ -0,0 +1,134 @@
+struct pinctrl_config {
+	/** Device address. */
+	uintptr_t reg;
+	/** List of states. */
+	const struct pinctrl_state **states;
+	/** Number of states. */
+	size_t states_cnt;
+};
+
+/**
+ * @brief Obtain the variable name storing pinctrl config for the given DT node
+ * identifier.
+ *
+ * @param node_id Node identifier.
+ */
+#define Z_PINCTRL_CONFIG_NAME(node_id) \
+	_CONCAT(__pinctrl_config, DEVICE_DT_NAME_GET(node_id))
+
+/**
+ * @brief Obtain the variable name storing pinctrl states for the given DT node
+ * identifier.
+ *
+ * @param node_id Node identifier.
+ */
+#define Z_PINCTRL_STATES_NAME(node_id) \
+	_CONCAT(__pinctrl_states, DEVICE_DT_NAME_GET(node_id))
+
+/**
+ * @brief Obtain the variable name storing the state information for the given
+ * state index and DT node identifier.
+ *
+ * @param state_idx State index.
+ * @param node_id Node identifier.
+ */
+
+/**
+ * @brief Helper macro to initialize the list of states.
+ *
+ * @param state_idx State index.
+ * @param node_id Node identifier.
+ */
+#define Z_PINCTRL_STATES_INIT(state_idx, node_id)			       \
+	COND_CODE_1(DT_PINCTRL_HAS_IDX(node_id, state_idx),		       \
+		    (&Z_PINCTRL_STATE_NAME(state_idx, node_id),), ())
+
+/**
+ * @brief Define all the states for the given node identifier.
+ *
+ * @param node_id Node identifier.
+ */
+#define Z_PINCTRL_STATES_DEFINE(node_id)				       \
+	UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),			       \
+		     Z_PINCTRL_STATE_DEFINE, node_id)			       \
+	static const struct pinctrl_state				       \
+	*Z_PINCTRL_STATES_NAME(node_id)[] = {				       \
+		UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),		       \
+			     Z_PINCTRL_STATES_INIT, node_id)		       \
+	};
+
+/**
+ * @brief Insert code depending on whether @p node_id has pinctrl entries (e.g.
+ * pinctrl-0) or not.
+ *
+ * @param node_id Node identifier.
+ * @param if_code Code inserted if node has pinctrl entries.
+ * @param else_code Code inserted if node does not have pinctrl entries.
+ */
+#define Z_PINCTRL_PRESENT_OR(node_id, if_code, else_code) \
+	COND_CODE_1(DT_PINCTRL_HAS_IDX(node_id, 0), if_code, else_code)
+
+/**
+ * @brief Define all pin control information for the given node identifier.
+ *
+ * This helper macro should be called together with device definition.
+ *
+ * @param node_id Node identifier.
+ */
+#define Z_PINCTRL_DEFINE(node_id)					       \
+	Z_PINCTRL_PRESENT_OR(node_id,					       \
+	(Z_PINCTRL_STATES_DEFINE(node_id)				       \
+	static const struct pinctrl_config Z_PINCTRL_CONFIG_NAME(node_id) = {  \
+		.reg = DT_REG_ADDR(node_id),				       \
+		.states = Z_PINCTRL_STATES_NAME(node_id),		       \
+		.states_cnt = ARRAY_SIZE(Z_PINCTRL_STATES_NAME(node_id)),      \
+	};), ())
+
+/**
+ * @brief Set the pinctrl state for the given device.
+ *
+ * @param dev Device instance.
+ * @param id Id of the state to be set.
+ * @return 0 on success, negative errno otherwise.
+ */
+int pinctrl_state_set(const struct device *dev, enum pinctrl_state_id id);
```

Comment:
We should implement this in a common way in terms of some "soc driver" specific implementation of `pinctrl_pin_configure`

Reply:
That probably can be used to avoid every implementation calling `pinctrl_lookup_state`, sounds reasonable. However, a public API for `pinctrl_pin_configure` would be problematic for the runtime case, since a pin configuration needs an associated state id.

Reply:
I agree that `pinctrl_pin_configure` shouldn't be used if you want driver state properly managed, but its a perfectly fine API if for example you want to init some pins at boot and don't need a full driver for them.

Reply:
Linux solves this problem using the "init" state, afaik. That means that the pinctrl driver would have an init function that sets all pins to the init state (if available). The same state would be applied after device de-init in the runtime case.

Reply:
There are some many cases people ask for wanting to tweak a pin for something and want a simple way to do that without a full blown driver.  We need some API to address this need that isn't associated w/a driver.

Reply:
Got it, I see a usecase for applications that use a certain peripheral out of the Zephyr driver model, directly in the app (e.g. using vendor HALs). I guess it's fine to have that, as soon as its purpose is clearly documented.

Reply:
I think we'll need a way to store arbitrary pinctrl configs in DT (not linked with a device) so that applications can retrieve them and use for runtime pinctrl or this custom case. b2401a0bf2328746625fedca36f26b8f28d20784 example app defines them directly in C code, which is cumbersome.

Reply:
> I think we'll need a way to store arbitrary pinctrl configs in DT (not linked with a device) so that applications can retrieve them and use for runtime pinctrl or this custom case. [b2401a0](https://github.com/zephyrproject-rtos/zephyr/commit/b2401a0bf2328746625fedca36f26b8f28d20784) example app defines them directly in C code, which is cumbersome.

We should see if `zephyr,user` can be used for this purpose.